File: app/src/main/java/com/rk/xededitor/BatchReplacement/BatchReplacement.java
package com.rk.xededitor.BatchReplacement;

import static com.rk.xededitor.MainActivity.Data.fragments;
import static com.rk.xededitor.MainActivity.Data.mTabLayout;
import static com.rk.xededitor.rkUtils.dpToPx;

import android.app.ProgressDialog;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.LinearLayout;

import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;

import com.rk.xededitor.BaseActivity;
import com.rk.xededitor.MainActivity.DynamicFragment;
import com.rk.xededitor.MainActivity.mAdapter;
import com.rk.xededitor.R;
import com.rk.xededitor.Settings.SettingsData;
import com.rk.xededitor.SimpleEditor.SimpleEditor;
import com.rk.xededitor.databinding.ActivityBatchReplacementBinding;
import com.rk.xededitor.rkUtils;

import java.util.Objects;
import java.util.Random;

import io.github.rosemoe.sora.widget.CodeEditor;

public class BatchReplacement extends BaseActivity {
  
  private ActivityBatchReplacementBinding binding;
  
  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    
    binding = ActivityBatchReplacementBinding.inflate(getLayoutInflater());
    setContentView(binding.getRoot());
    
    Toolbar toolbar = binding.toolbar;
    setSupportActionBar(toolbar);
    Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
    getSupportActionBar().setTitle("Replace");
    
    if (SettingsData.isDarkMode(this) && SettingsData.isOled(this)) {
      findViewById(R.id.drawer_layout).setBackgroundColor(Color.BLACK);
      findViewById(R.id.appbar).setBackgroundColor(Color.BLACK);
      findViewById(R.id.mainBody).setBackgroundColor(Color.BLACK);
      findViewById(R.id.toolbar).setBackgroundColor(Color.BLACK);
      getWindow().setNavigationBarColor(Color.BLACK);
      Window window = getWindow();
      window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
      window.setStatusBarColor(Color.BLACK);
    }
    
    
  }
  
  private View newEditBox(String hint) {
    LinearLayout rootLinearLayout = new LinearLayout(this);
    rootLinearLayout.setLayoutParams(new LinearLayout.LayoutParams(
        ViewGroup.LayoutParams.MATCH_PARENT,
        ViewGroup.LayoutParams.WRAP_CONTENT));
    rootLinearLayout.setTag("keyRep");
    rootLinearLayout.setOrientation(LinearLayout.VERTICAL);
    
    // Create the inner LinearLayout
    LinearLayout innerLinearLayout = new LinearLayout(this);
    LinearLayout.LayoutParams innerParams = new LinearLayout.LayoutParams(
        ViewGroup.LayoutParams.MATCH_PARENT,
        dpToPx(50, this));  // height is 50dp
    innerParams.setMargins(dpToPx(22, this), dpToPx(10, this), dpToPx(22, this), 0);
    innerLinearLayout.setLayoutParams(innerParams);
    innerLinearLayout.setTag("keyword");
    
    // Set the background for innerLinearLayout
    Drawable drawable = ContextCompat.getDrawable(this, R.drawable.edittext);
    innerLinearLayout.setBackground(drawable);
    
    // Create the EditText
    EditText editText = new EditText(this);
    LinearLayout.LayoutParams editTextParams = new LinearLayout.LayoutParams(
        ViewGroup.LayoutParams.MATCH_PARENT,
        ViewGroup.LayoutParams.MATCH_PARENT);
    editText.setLayoutParams(editTextParams);
    editText.setPadding(dpToPx(8, this), 0, dpToPx(5, this), 0);  // paddingStart 8dp, paddingEnd 5dp
    editText.setId(View.generateViewId());
    editText.setSingleLine(true);
    editText.setHint(hint);
    editText.setBackgroundColor(android.graphics.Color.TRANSPARENT);
    
    // Add EditText to inner LinearLayout
    innerLinearLayout.addView(editText);
    
    rootLinearLayout.addView(innerLinearLayout);
    return rootLinearLayout;
  }
  
  public void addBatch(View v) {
    int random_int = new Random().nextInt();
    ((LinearLayout) findViewById(R.id.mainBody)).addView(newEditBox("keyword (Regex)"));
    ((LinearLayout) findViewById(R.id.mainBody)).addView(newEditBox("replacement"));
    View view = new View(this);
    
    // Set the width and height of the View
    LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
        dpToPx(0, this), // Width in dp
        dpToPx(20, this)   // Height in dp
    );
    view.setLayoutParams(params);
    ((LinearLayout) findViewById(R.id.mainBody)).addView(view);
    
    findViewById(R.id.removeBatch).setVisibility(View.VISIBLE);
  }
  
  public void removeBatch(View v) {
    LinearLayout linearLayout = findViewById(R.id.mainBody);
    int childCount = linearLayout.getChildCount();
    if (childCount > 3) {
      // Remove the last child first
      linearLayout.removeViewAt(childCount - 1);
      // Remove the second-to-last child
      linearLayout.removeViewAt(childCount - 2);
      // Remove the third-to-last child
      linearLayout.removeViewAt(childCount - 3);
      if (linearLayout.getChildCount() <= 3) {
        findViewById(R.id.removeBatch).setVisibility(View.GONE);
      }
    } else {
      findViewById(R.id.removeBatch).setVisibility(View.GONE);
    }
    
  }
  
  public void replace_all(View v) {
    ProgressDialog dialog = ProgressDialog.show(this, "",
        "Please wait...", true);
    LinearLayout linearLayout = findViewById(R.id.mainBody);
    for (int i = 0; i < linearLayout.getChildCount(); i++) {
      View view = linearLayout.getChildAt(i);
      if (view instanceof LinearLayout) {
        LinearLayout l = (LinearLayout) ((LinearLayout) view).getChildAt(0);
        EditText editText = (EditText) l.getChildAt(0);
        if (editText.getHint().equals("keyword (Regex)")) {
          View viewx = linearLayout.getChildAt(i + 1);
          LinearLayout lx = (LinearLayout) ((LinearLayout) viewx).getChildAt(0);
          EditText editTextx = (EditText) lx.getChildAt(0);
          String keyword = editText.getText().toString();
          String replacement = editTextx.getText().toString();
          
          if (fragments != null){
           var editor = mAdapter.getCurrentEditor();
           editor.setText(editor.getText().toString().replaceAll(keyword, replacement));
            
          }else if(getIntent().getExtras().getBoolean("isExt",false)){
            SimpleEditor.editor.setText(SimpleEditor.editor.getText().toString().replaceAll(keyword, replacement));
          }
          
          
          
          
        }
      }
    }
    dialog.hide();
    rkUtils.toast(this, "Action Completed");
  }
  
  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    // Handle action bar item clicks here
    int id = item.getItemId();
    if (id == android.R.id.home) {
      // Handle the back arrow click here
      onBackPressed();
      return true;
    }
    return super.onOptionsItemSelected(item);
  }
}

File: app/src/main/java/com/rk/xededitor/MainActivity/BatchReplacement.kt
package com.rk.xededitor.MainActivity

import android.graphics.Color
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import com.rk.libcommons.LoadingPopup
import com.rk.xededitor.BaseActivity
import com.rk.xededitor.MainActivity.tabs.editor.EditorFragment
import com.rk.xededitor.MainActivity.tabs.core.FragmentType
import com.rk.xededitor.R
import com.rk.xededitor.SimpleEditor.SimpleEditor
import com.rk.xededitor.databinding.ActivityBatchReplacementBinding
import com.rk.xededitor.rkUtils.dpToPx
import com.rk.xededitor.rkUtils.toast

/*
 *
 * DO NOT TRANSLATE
 *
 * */

class BatchReplacement : BaseActivity() {
    private lateinit var binding: ActivityBatchReplacementBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityBatchReplacementBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val toolbar = binding.toolbar
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = resources.getString(R.string.title_activity_batch_replacement)
    }

    private fun newEditBox(hint: String): View {
        val rootLinearLayout = LinearLayout(this)
        rootLinearLayout.layoutParams =
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            )
        rootLinearLayout.tag = "keyRep"
        rootLinearLayout.orientation = LinearLayout.VERTICAL

        // Create the inner LinearLayout
        val innerLinearLayout = LinearLayout(this)
        val innerParams =
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                dpToPx(50f, this),
            ) // height is 50dp
        innerParams.setMargins(dpToPx(22f, this), dpToPx(10f, this), dpToPx(22f, this), 0)
        innerLinearLayout.layoutParams = innerParams
        innerLinearLayout.tag = "keyword"

        // Set the background for innerLinearLayout
        val drawable = ContextCompat.getDrawable(this, R.drawable.edittext)
        innerLinearLayout.background = drawable

        // Create the EditText
        val editText = EditText(this)
        val editTextParams =
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
        editText.layoutParams = editTextParams
        editText.setPadding(
            dpToPx(8f, this),
            0,
            dpToPx(5f, this),
            0,
        ) // paddingStart 8dp, paddingEnd 5dp
        editText.id = View.generateViewId()
        editText.isSingleLine = true
        editText.hint = hint
        editText.setBackgroundColor(Color.TRANSPARENT)

        // Add EditText to inner LinearLayout
        innerLinearLayout.addView(editText)

        rootLinearLayout.addView(innerLinearLayout)
        return rootLinearLayout
    }

    fun addBatch(v: View?) {
        (findViewById<View>(R.id.mainBody) as LinearLayout).addView(newEditBox("keyword (Regex)"))
        (findViewById<View>(R.id.mainBody) as LinearLayout).addView(newEditBox("replacement"))
        val view = View(this)

        // Set the width and height of the View
        val params =
            LinearLayout.LayoutParams(
                dpToPx(0f, this), // Width in dp
                dpToPx(20f, this), // Height in dp
            )
        view.layoutParams = params
        (findViewById<View>(R.id.mainBody) as LinearLayout).addView(view)

        findViewById<View>(R.id.removeBatch).visibility = View.VISIBLE
    }

    fun removeBatch(v: View?) {
        val linearLayout = findViewById<LinearLayout>(R.id.mainBody)
        val childCount = linearLayout.childCount
        if (childCount > 3) {
            // Remove the last child first
            linearLayout.removeViewAt(childCount - 1)
            // Remove the second-to-last child
            linearLayout.removeViewAt(childCount - 2)
            // Remove the third-to-last child
            linearLayout.removeViewAt(childCount - 3)
            if (linearLayout.childCount <= 3) {
                findViewById<View>(R.id.removeBatch).visibility = View.GONE
            }
        } else {
            findViewById<View>(R.id.removeBatch).visibility = View.GONE
        }
    }

    fun replaceAll(v: View?) {
        LoadingPopup(this, 100L)
        val linearLayout = findViewById<LinearLayout>(R.id.mainBody)
        for (i in 0 until linearLayout.childCount) {
            val view = linearLayout.getChildAt(i)
            if (view is LinearLayout) {
                val l = view.getChildAt(0) as LinearLayout
                val editText = l.getChildAt(0) as EditText
                if (editText.hint == "keyword (Regex)") {
                    val viewx = linearLayout.getChildAt(i + 1)
                    val lx = (viewx as LinearLayout).getChildAt(0) as LinearLayout
                    val editTextx = lx.getChildAt(0) as EditText
                    val keyword = editText.text.toString()
                    val replacement = editTextx.text.toString()

                    if (MainActivity.activityRef.get() != null) {
                        MainActivity.activityRef.get()?.adapter?.getCurrentFragment()?.let {
                            if (it.type!! == FragmentType.EDITOR){
                                (it.fragment as EditorFragment).editor?.setText((it.fragment as EditorFragment).editor?.text.toString().replace(keyword, replacement))
                            }else{
                                throw RuntimeException("Unsupported Fragment type")
                            }
                        }
                    } else if (intent.extras?.getBoolean("isExt", false) == true) {
                        // if we are working with external editor
                        SimpleEditor.editor?.setText(
                            SimpleEditor.editor!!.text.toString().replace(keyword, replacement)
                        )
                    }
                }
            }
        }

        toast(getString(R.string.action_done))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here
        val id = item.itemId
        if (id == android.R.id.home) {
            // Handle the back arrow click here
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}
